New: Update DB to store all releases for an album

ta264 opened the pull request:
#### Database Migration
YES

#### Description
Rework schema as discussed.  Separates out AlbumReleases into their own table, and separates the Artists table into ArtistMetadata and Artist.  Albums now point to ArtistMetadata instead of Artist and we also link tracks to their actual artist instead of the album artist.  We then store tracks for every release instead of just the  current release.

You can view a model of the schema [here].(https://app.sqldbm.com/MySQL/Share/g5NOxqRjfjKjcpQyRkafM0GFrngIE8md_DYjF4jNYw0)

In the process I have removed the concept of a monitored track since that made life easier.  Monitored status is now associated with Artists and Albums (as it is in the UI). Releases have a monitored status as prep for supporting multiple releases, though the code constrains it to have exactly one monitored release at any one time. 

I have also implemented automatic release switching which can be configured by album. It defaults to true, and will be set to true in the migration. 

I have set up the migration to create a dummy release for each album.  It then edits the last sync time and last run time of the refreshartists task to trigger a full metadata refresh, which pulls in all the new releases and tracks.

This needs to be run against the metadata server here:
https://github.com/lidarr/LidarrAPI.Metadata/pull/80

I think I have fixed all the tests that should be fixed.  There are two failing artist search ones ('Where's Kitty?' and 'The Academy Is...' but they fail for me in develop too.

#### Todos
- [x] Tests
- [x] Documentation
- [x] Load data for artists by track
- [x] migrate libraries from old to new schema


#### Issues Fixed or Closed by this PR

*

Qstick commented on src/NzbDrone.Common/Cloud/LidarrCloudRequestBuilder.cs:
```diff
@@ -16,7 +16,7 @@ public LidarrCloudRequestBuilder()
                 .CreateFactory();
 
-            Search = new HttpRequestBuilder("https://api.lidarr.audio/api/v0.3/{route}")
+            Search = new HttpRequestBuilder("http://localhost:5001/{route}")
```
This would be bumped to 0.4 when metadata changes are merged. You can set a custom metadata server in settings FYI, instead of changing it in code.

ta264 replied:
Sorry, didn't mean to commit that. Will remove it.

ta264 commented on src/NzbDrone.Core/Datastore/Migration/023_add_release_groups_etc.cs:
```diff
@@ -0,0 +1,105 @@
+            Delete.Table("Artists");
+            Delete.Table("Albums");
+            Delete.Table("Tracks");
+            Delete.Table("TrackFiles");
```
Yeah, now I've started doing the actual migration I've done it how you suggest - editing the existing tables.

At the moment there is still ReleaseGroup (which has extra information) and Album. I've done the minimal renaming to make the new schema work, so mostly on adding new items where the extra info is required. ReleaseGroup is then implicitly cast to Album in the rest of the code to avoid messing with it. Happy to try a different approach if you think it would be better. I guess I could try augmenting the album class rather than creating a new one but was worried I'd get myself in a mess around when the extra info is populated or not

ta264 replied:
Thinking about this some more, I think you're probably right that I could merge the ReleaseGroup class into Album, retaining the old properties by calculating from the new ones. That would probably be easier than casting to Album. I'll have a go.

Qstick replied:
Sounds good, we could probably do this in stages to if you want to break it up. Something like:

- break out releases into table
- add artist metadata table
- etc.

Up to you on that though

ta264 commented:
Still need to fix up the tests but thought you might like to see what I have so far.

ta264 commented:
I think this is pretty much ready to go, or at least ready for you guys to hit it with the big red pen...

Qstick commented:
Sounds good, I'll spin it up and start looking thru the code

ta264 commented:
I've reworked a bit as you suggested. The database now supports monitoring multiple releases (though the code enforces only one at a time for now). Also added an 'AnyReleaseOk' attribute to albums and implemented automatic release switching.

Qstick commented:
In testing the migration only brings over the current release as you mentioned, thus if something is downloaded prior to a refresh happening there is only the one release for Lidarr to choose from. I suppose this isn't the end of the world given that's pretty much how things work now.

Qstick commented on src/NzbDrone.Core/Datastore/Migration/023_add_release_groups_etc.cs:
```diff
@@ -0,0 +1,231 @@
+
+            // RELEASES TABLE - Do this before we mess with the Albums table
+
+            Create.TableForModel("Releases")
```
This table name could possibly get confused with indexer releases, to which we have an API endpoint module and resource named similar. Should we use AlbumReleases?

ta264 replied:
I've renamed to Album Release locally. Afraid I got distracted by the parsing we were talking about so forgot to push. Will do it tomorrow.

ta264 commented:
The trouble is even if you stick in the rest of the releases you don't have the track data for the other releases until the refresh finishes, so I think all you gain is the list of releases in the drop down. But changing that no longer triggers a metadata refresh so wasn't sure there was much point.

Qstick commented on src/Lidarr.Api.V1/Albums/AlbumReleaseResource.cs:
```diff
@@ -7,15 +7,31 @@ namespace Lidarr.Api.V1.Albums
+        public int AlbumId { get; set; }
+        public string ForeignReleaseId { get; set; }
         public string Title { get; set; }
-        public DateTime? ReleaseDate { get; set; }
```
ex. <https://musicbrainz.org/release-group/1346c819-d57b-4c38-aa55-1e599dfc3645>

ta264 replied:
On `develop` we only have release dates for the release group and not the release, which I have kept in this PR, so I removed the field from the `AlbumReleaseResource` because it's not used and it would just be returning the date from the `Album` anyway.  I can update the metadata to fetch the release date for each `AlbumRelease` if you want.

ta264 replied:
I agree. I'll have a go at sorting this out. Might rename the the Album release date as OriginalYear since that's what it is

Qstick commented on src/NzbDrone.Core/Extras/Metadata/Consumers/Wdtv/WdtvMetadata.cs:
```diff
@@ -111,10 +111,10 @@ public override MetadataFileResult TrackMetadata(Artist artist, TrackFile trackF
                     details.Add(new XElement("id", artist.Id));
                     details.Add(new XElement("title", string.Format("{0} - {1} - {2}", artist.Name, track.TrackNumber, track.Title)));
-                    details.Add(new XElement("artist_name", artist.Name));
+                    details.Add(new XElement("artist_Name", artist.Metadata.Value.Name));
```
Did you mean to change case on `artist_Name`

ta264 replied:
Oops...

Qstick commented on src/NzbDrone.SignalR/NzbDronePersistentConnection.cs:
```diff
@@ -31,17 +31,6 @@ public NzbDronePersistentConnection(IConfigFileProvider configFileProvider)
-            }
-
-            _messageHistory[message.Name] = message.Body.ToJson();
-
```
What was this removed for?

ta264 replied:
Good spot - I'll put it back.  I removed this when I was tracking down a bug that turned out to be caused by `ToJson` evaluating all the `LazyLoaded` instances causing an infinite loop.

Qstick commented on frontend/src/Album/Details/TrackRow.js:
```diff
@@ -178,7 +178,6 @@ TrackRow.propTypes = {
   id: PropTypes.number.isRequired,
   albumId: PropTypes.number.isRequired,
   trackFileId: PropTypes.number,
-  monitored: PropTypes.bool.isRequired,
```
Should we keep in track monitor field in case we decide to support selective track saving (only keep certain tracks from a album), or support a service that downloads by track?

ta264 replied:
As with the `Album` tags, my view is that we should remove it because it adds complication and is likely to break given it's not in use.  It can always be reinstated when it's needed.  Again, happy to leave it if you prefer.

Qstick replied:
I’m good with that

Qstick commented on src/Lidarr.Api.V1/Tracks/TrackModuleWithSignalR.cs:
```diff
@@ -87,9 +87,7 @@ protected List<TrackResource> MapToResource(List<Track> tracks, bool includeArti
-
-                    var series = track.Artist ?? artistDict.GetValueOrDefault(tracks[i].ArtistId) ?? _artistService.GetArtist(tracks[i].ArtistId);
-                    artistDict[series.Id] = series;
+                    var series = track.Artist.Value;
```
Last thing, Can we fix series reference while we are here :)

ta264 commented on src/NzbDrone.Core/Datastore/Migration/023_add_release_groups_etc.cs:
```diff
@@ -0,0 +1,246 @@
+                .Column("Releases")
+                .Column("CurrentRelease")
+                .Column("LastDiskSync")
+                .Column("Overview")
```
I could hook this up using the wikidata links like the artist overview if you like.  See our favourite release [here](https://musicbrainz.org/release-group/1346c819-d57b-4c38-aa55-1e599dfc3645).

ta264 commented on src/NzbDrone.Core/Datastore/Migration/023_add_release_groups_etc.cs:
```diff
@@ -0,0 +1,246 @@
+                .Column("TitleSlug")
+                .Column("Label")
+                .Column("SortTitle")
+                .Column("Tags")
```
My view is that stuff that isn't used/properly tested should be removed and can always be put back in when the album-centric actually happens.  Otherwise it might accidentally get used when it doesn't work properly.  But happy to leave it if you guys prefer.

Qstick replied:
I’m good adding back when used.

ta264 commented:
I'll sort out the artist metadata service so it handles Artist/Track deleted events.  The only argument against is that it creates a whole bunch of database queries every time a track is deleted so it might hurt performance a bit when it doesn't matter if the metadata hangs around for a bit, but I imagine the impact would be small.
